content/multi-source-extension-overloads.md

```diff
@@ -0,0 +1,178 @@
+---
+layout: sip
+permalink: /sips/:title.html
+stage: implementation
+status: waiting-for-implementation
+title: SIP-54 - Multi-Source Extension Overloads
+---
+
+**By: Sébastien Doeraene and Martin Odersky**
+
+## History
+
+| Date          | Version            |
+|---------------|--------------------|
+| Mar 10th 2023 | Initial Draft      |
+
+## Summary
+
+We propose to allow overload resolution of `extension` methods with the same name but imported from several sources.
+For example, given the following definitions:
+
+```scala
+class Foo
+class Bar
+
+object A:
+  extension (foo: Foo) def meth(): Foo = foo
+  def normalMeth(foo: Foo): Foo = foo
+
+object B:
+  extension (bar: Bar) def meth(): Bar = bar
+  def normalMeth(bar: Bar): Bar = bar
+```
+
+and the following use site:
+
+```scala
+import A.*
+import B.*
+
+val foo: Foo = ???
+foo.meth() // works with this SIP; "ambiguous import" without it
+
+// unchanged:
+meth(foo)() // always ambiguous, just like
+normalMeth(foo) // always ambiguous
+```
+
+## Motivation
+
+Extension methods are a great, straightforward way to extend external classes with additional methods.
+One classical example is to add a `/` operation to `Path`:
+
+```scala
+import java.nio.file.*
+
+object PathExtensions:
+  extension (path: Path)
+    def /(child: String): Path = path.resolve(child).nn
+
+def app1(): Unit =
+  import PathExtensions.*
+  val projectDir = Paths.get(".") / "project"
+```
+
+However, as currently specified, they do not compose, and effectively live in a single flat namespace.
+This is understandable from the spec--the *mechanism**, which says that they are just regular methods, but is problematic from an intuitive point of view--the *intent*.
+
+For example, if we also use another extension that provides `/` for `URI`s, we can use it in a separate scope as follows:
+
+```scala
+import java.net.URI
+
+object URIExtensions:
+  extension (uri: URI)
+    def /(child: String): URI = uri.resolve(child)
+
+def app2(): Unit =
+  import URIExtensions.*
+  val rootURI = new URI("https://www.example.com/")
+  val projectURI = rootURI / "project/"
+```
+
+The above does not work anymore if we need to use *both* extensions in the same scope.
+The code below does not compile:
+
+```scala
+def app(): Unit =
+  import PathExtensions.*
+  import URIExtensions.*
+
+  val projectDir = Paths.get(".") / "project"
+  val rootURI = new URI("https://www.example.com/")
+  val projectURI = rootURI / "project/"
+  println(s"$projectDir -> $projectURI")
+end app
+```
+
+*Both* attempts to use `/` result in error messages of the form
+
+```
+Reference to / is ambiguous,
+it is both imported by import PathExtensions._
+and imported subsequently by import URIExtensions._
+```
+
+### Workarounds
```

Comment:
Just for the completeness of the SIP - there seems to be one more workaround, probably first (partially) mentioned in https://github.com/lampepfl/dotty/issues/13613. The trick is to define extension methods in given instances of type `{}` (so that they don't clash with other useful givens during implicit search). The given instances should have different names so that they could be imported to the scope where we need to use the extension methods. E.g.
```scala
class Foo
class Bar
object A:
  given fooOps: {} with
    extension (foo: Foo) def meth(): Foo = foo
object B:
  given fooOps: {} with
    extension (bar: Bar) def meth(): Bar = bar
```

```scala
import A.given
import B.given
val foo: Foo = ???
val bar: Bar = ???
val a = foo.meth()
val b = bar.meth()
```